funsor/sum_product.py

```diff
@@ -45,6 +45,61 @@ def _partition(terms, sum_vars):
     return components
 
 
+def _unroll_plate(factors, plate, size, var_to_ordinal):
+    # replicated variables
+    plate_vars = set()
+    for var, ordinal in var_to_ordinal.items():
+        if plate in ordinal:
+            plate_vars.add(var)
+
+    # unroll factors
+    unrolled_factors = []
+    for factor in factors:
+        if plate in factor.inputs:
+            unrolled_factors.extend([factor(
+                    **{plate: i},
+                    **{var: "{}_{}{}".format(var, plate, i) for var in plate_vars}
+                ) for i in range(size)])
+        else:
+            unrolled_factors.append(factor)
+
+    # unroll variables
+    for var in plate_vars:
+        ordinal = var_to_ordinal.pop(var)
+        var_to_ordinal.update({"{}_{}{}".format(var, plate, i): ordinal.difference(plate)
+                               for i in range(size)})
+
+    return unrolled_factors, var_to_ordinal
+
+
+def partial_unroll(factors, plates=frozenset()):
```

Comment:
I think it would make sense for `partial_unroll` to take an `eliminate` argument and return an updated version of `eliminate` with any plates removed and variables replaced with unrolled versions.